arrows/core/derive_metadata.cxx

```diff
@@ -210,47 +237,86 @@ double
 compute_gsd( kwiver::vital::metadata_sptr const& metadata,
              size_t frame_width, size_t frame_height )
 {
-  if ( frame_width == 0 || frame_height == 0 )
+  if( frame_width < 1 || frame_height < 1 )
   {
-    VITAL_THROW( kv::invalid_value, "frame dimensions cannot be zero" );
+    VITAL_THROW( kv::invalid_value, "frame dimensions must both be positive" );
   }
 
-  kv::vector_2d sensor_fov;
+  // RP1201 method
   try
   {
-    // Attempt to acquire sensor FOV
-    sensor_fov = get_sensor_fov( metadata );
+    // Intermediate values
+    kv::rotation_d const total_rotation = get_total_rotation( metadata );
+    double yaw, pitch, roll;
+    total_rotation.get_yaw_pitch_roll( yaw, pitch, roll );
+    // Ideally slant range should be pre-calculated
+    double const slant_range = get_slant_range( metadata );
+    double const sensor_horizontal_fov_rad = get_sensor_horizontal_fov( metadata );
+    double const sensor_vertical_fov_rad = get_sensor_vertical_fov( metadata );
+
+    if( slant_range > 0 && slant_range <= 5000000 &&
+        sensor_horizontal_fov_rad >= 0 &&
+        sensor_horizontal_fov_rad <= kv::pi &&
+        sensor_vertical_fov_rad >= 0 && sensor_vertical_fov_rad <= kv::pi )
+    {
+
+      // Approximate dimensions of image on ground plane
+      double const gsd_horiz =
+        2.0 * slant_range * std::tan( ( sensor_horizontal_fov_rad / 2.0 ) /
+                                      static_cast< float >( frame_width ) );
+      double const gsd_vert =
+        2.0 * slant_range * std::tan(
+          ( sensor_vertical_fov_rad / 2.0 ) /
+          static_cast< float >( frame_height ) ) /
+        std::sin( -pitch );
+
+      // GSD is the geometric mean of each dimensions's GSD
+      // All values in meters per pixel
+      return std::sqrt( gsd_horiz * gsd_vert );
+    }
+  }
+  catch ( kv::invalid_value const& e )
+  {
+    // Move onto the next case
+  }
+
+  // RP 1201 horizontal axis only
+  try
+  {
+    auto const horizontal_sensor_fov = get_sensor_horizontal_fov( metadata );
+    // Note that the reference implementation doesn't use computed slant range for
+    // this method
+    auto const slant_range = get_slant_range( metadata );
+
+    if( slant_range > 0 && slant_range < 5000000 &&
+        horizontal_sensor_fov >= 0 && horizontal_sensor_fov <= 180 )
+    {
+      return 2000.0 * slant_range * tan(
```

Comment:
why is this case times 2000 and not 2?

Reply:
it would be better to refactor this to avoid having duplicate code for computing the horizontal gsd.  If there is a bug in the algorithm, we only want to need to remember to fix it one place.

Reply:
This is indeed a bug. And it was because they're actually reporting GSD in millimetres in the reference code.  

I've refactored to have a single correct function.